File: Go4Lunch/app/src/main/java/com/openclassrooms/go4lunch/data/local/db/AppDatabase.java
package com.openclassrooms.go4lunch.data.local.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.openclassrooms.go4lunch.data.local.db.dao.PlaceDao;
import com.openclassrooms.go4lunch.data.local.db.dao.UserDao;
import com.openclassrooms.go4lunch.data.model.db.Place;
import com.openclassrooms.go4lunch.data.model.db.User;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Room Database Class for application
 */
@Database(entities = {Place.class, User.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "go4lunch.db";
    private static AppDatabase instance;

    // For Singleton instantiation
    private static final Object LOCK = new Object();

    public abstract PlaceDao placeDao();

    public abstract UserDao userDao();

    public static AppDatabase database(Context context) {
        synchronized (LOCK) {
            if (instance == null) {
                if (!isRunningInstrumentedTest()) {
                    instance = Room.databaseBuilder(context, AppDatabase.class, DATABASE_NAME)
                            .build();
                }

                if (isRunningInstrumentedTest()) {
                    instance = Room.inMemoryDatabaseBuilder(context, AppDatabase.class)
                            .allowMainThreadQueries()
                            .build();
                }
            }
            return instance;
        }
    }

    private static AtomicBoolean isRunningInstrumentedTest;

    public static synchronized boolean isRunningInstrumentedTest() {
        if (null == isRunningInstrumentedTest) {
            boolean istest;

            try {
                Class.forName("androidx.test.espresso.Espresso");
                istest = true;
            } catch (ClassNotFoundException e) {
                istest = false;
            }
            isRunningInstrumentedTest = new AtomicBoolean(istest);
        }
        return isRunningInstrumentedTest.get();
    }
}

File: RealEstateManager/app/src/main/java/com/openclassrooms/realestatemanager/data/cache/AppDatabase.kt
package com.openclassrooms.realestatemanager.data.cache

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import com.openclassrooms.realestatemanager.data.cache.dao.PhotoDao
import com.openclassrooms.realestatemanager.data.cache.dao.PropertyDao
import com.openclassrooms.realestatemanager.models.property.*

@Database(entities = [Property::class, Photo::class], version = 1, exportSchema = false)
@TypeConverters(*arrayOf(
    PropertyTypeConverter::class,
        PropertyStatusConverter::class,
        PhotoTypeConverter::class,
        InterestPointConverter::class,
        DateConverter::class
))
abstract class AppDatabase : RoomDatabase() {

    abstract fun propertyDao(): PropertyDao
    abstract fun photoDao(): PhotoDao

    companion object {
        const val DATABASE_NAME: String = "real_estate_manger.db"
    }
}